opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,595 @@
+}
+
+template<typename ActionT>
+void FollowingServer::getPreemptedGoalIfRequested(
+  typename std::shared_ptr<const typename ActionT::Goal> goal,
+  const std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server)
+{
+  if (action_server->is_preempt_requested()) {
+    goal = action_server->accept_pending_goal();
+  }
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfCancelled(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_cancel_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was cancelled. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfPreempted(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_preempt_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was preempted. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+void FollowingServer::followObject()
+{
+  std::lock_guard<std::mutex> lock(*dynamic_params_lock_);
+  action_start_time_ = this->now();
+  rclcpp::Rate loop_rate(controller_frequency_);
+
+  auto goal = following_action_server_->get_current_goal();
+  auto result = std::make_shared<FollowObject::Result>();
+
+  if (!following_action_server_ || !following_action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server unavailable or inactive. Stopping.");
+    return;
+  }
+
+  if (checkAndWarnIfCancelled(following_action_server_, "follow_object")) {
+    following_action_server_->terminate_all();
+    return;
+  }
+
+  getPreemptedGoalIfRequested(goal, following_action_server_);
+  num_retries_ = 0;
+
+  try {
+    RCLCPP_INFO(
+      get_logger(),
+      "Attempting to follow object at position (%0.2f, %0.2f).",
+      goal->object_pose.pose.position.x, goal->object_pose.pose.position.y);
+
+    // Construct initial estimate of where the object is located in fixed_frame
+    auto object_pose = goal->object_pose;
+    object_pose.header.stamp = rclcpp::Time(0);
+    tf2_buffer_->transform(object_pose, object_pose, fixed_frame_);
+
+    // Get initial detection of the object before proceeding to move
+    doInitialPerception(object_pose);
+    RCLCPP_INFO(get_logger(), "Successful initial object detection");
+
+    // Following control loop: while not timeout, run controller
+    auto start = this->now();
+    rclcpp::Duration max_duration = goal->max_duration;
+    while (rclcpp::ok()) {
+      try {
+        // Check if we have run out of time
+        if (this->now() - start > max_duration && max_duration.seconds() > 0.0) {
+          RCLCPP_INFO(get_logger(), "Exceeded max duration. Stopping.");
+          result->total_elapsed_time = this->now() - action_start_time_;
+          result->error_code = FollowObject::Result::TIMEOUT;
```

Comment:
For successful completion, we don't set error codes. Error codes are due to failures. So this should be removed and probably so should be that enum value in the action definition)